azurelinuxagent/common/protocol/wire.py

```diff
@@ -908,49 +918,59 @@ def _call_hostplugin_with_container_check(self, host_func):
-        # they will not be refreshed even if a goal state refresh is called before retrying the host_func.
+    def send_request_using_appropriate_channel(self, direct_func, host_func):
+        """
+        Determines which communication channel to use. By default, the primary channel is direct, host channel is secondary.
+        We call the primary channel first and return on success. If primary fails, we try secondary. If secondary fails,
+        we return and *don't* switch the default channel. If secondary succeeds, we change the default channel.
+        This method doesn't raise since the calls to direct_func and host_func are already wrapped and handle any exceptions.
+        Possible return values are manifest, artifacts profile, boolean or None.
+        """
+        direct_channel = lambda: self.__send_request_using_direct_channel(direct_func)
+        host_channel = lambda: self.__send_request_using_host_channel(host_func)
 
-        if not HostPluginProtocol.is_default_channel():
-            ret = None
-            try:
-                ret = direct_func()
-
-                # Different direct channel functions report failure in different ways: by returning None, False,
-                # or raising ResourceGone or InvalidContainer exceptions.
-                if not ret:
-                    logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
-                                                            "switching to host plugin.")
-            except (ResourceGoneError, InvalidContainerError) as e:  # pylint: disable=C0103
-                logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
-                                                        "switching to host plugin. Error: {0}".format(ustr(e)))
-
-            if ret:
-                return ret
+        if HostPluginProtocol.is_default_channel:
+            primary_channel, secondary_channel = host_channel, direct_channel
         else:
-            logger.periodic_info(logger.EVERY_HALF_DAY, "[PERIODIC] Using host plugin as default channel.")
+            primary_channel, secondary_channel = direct_channel, host_channel
 
-        ret = self._call_hostplugin_with_container_check(host_func)
+        ret = primary_channel()
+        if ret is not None:
+            return ret
 
-        if not HostPluginProtocol.is_default_channel():
-            logger.info("Setting host plugin as default channel from now on. "
-                        "Restart the agent to reset the default channel.")
-            HostPluginProtocol.set_default_channel(True)
+        ret = secondary_channel()
+        if ret is not None:
+            HostPluginProtocol.is_default_channel = not HostPluginProtocol.is_default_channel
+            logger.info("Default channel changed to {0}.", "host" if HostPluginProtocol.is_default_channel else "direct")
```

Comment:
How verbose can this be? Should we also send a Kusto event for this? We would have more data to check how frequently we switch channels that way

Reply:
So this only happens if the primary fails **and** secondary succeeds. It will happen again only if that same channel we just set as default fails again **and** the other now succeeds.

I don't think this happens often, usually just one channel is down so the other takes over, in which case we wouldn't be hitting this successively. The only scenario where this might be verbose is if both channels are flaky **at the same time**, which I believe is a very small chance.

I like the idea of logging it to Kusto, if at least just for this release to give us a sense of how often it happens. Let me create a new operation for this.